### DISCUSSION TITLE
RawHid + Gamepad + Keyboard

### DISCUSSION THREAD

**Discussion by DanForever** [NicoHood/HID#269] [General]
I'm wondering why RawHid can't be used in combination with any other device type. With my Teensy 4.0, I've got RawHid + Keyboard working using the teensy libraries in single report mode.

With my Pro Micro, using NicoHood/HID, I can't combine RawHid with anything else, and I'm wondering why? What does this library do differently from Teensy's? Is it a hardware limitation?

Also, on the subject of multireports, I have a Logitech G700s mouse which uses a combination of RawHid + mouse using multireports (so I know that's possible). At least, I'm pretty sure it's raw hid that's used for their "HID++" protocol.

**Comment by NicoHood** [comment]
I dont know anymore. I've disabled it for some good reasons, mainly because it had problems and I aint got time to make it work again. Of course it is possible. So feel free to get it working and open a PR

**Reply by DanForever** [reply]
OK, challenge accepted I guess :D

Do you remember, or have any insight why you were having problems or what the cause might have been?

**Reply by NicoHood** [reply]
The issue was that sending exactly the EP size (64) or 1 byte less or more was a problem. Also if you wait to long to read it will block the host I think. Something like that. Also I think it was only working in single report because of that blocking, not in multi report. You can try to reenable it, I think it currently is disabled. Also there are some opened issue where people discussed the problems, so you might want to check that out.

I cross my fingers for you! Would be great to implement that!

**Comment by DanForever** [comment]
So I've written some code that prints out some debug information about the `PluggableUSBModule` we're in.

If I initialize _only_ RawHID, (`RawHID.begin( buffer, bufferSize );` I get output like so:
```Debug RawHid
RawHid Configuration:
 - Endpoint: 4
 - Interface: 2
 - Endpoint Count: 1
 - Interface Count: 1
```
If I also enable GamePad _afterwards_ (`Gamepad1.begin();`), then I get this:
```Debug RawHid
RawHid Configuration:
 - Endpoint: 0
 - Interface: 0
 - Endpoint Count: 1
 - Interface Count: 1
Debug Gamepad
Gamepad Configuration:
 - Endpoint: 4
 - Interface: 2
 - Endpoint Count: 1
 - Interface Count: 1
```

What is currently causing the RawHID interface to be disabled? The debug info about RawHID is dumped before we get to Gamepad, so I don't really understand what's going on here. I would have thought, if anything, that maybe the gamepad overrides RawHID once it's created, or Gamepad initialization fails because we've already got RawHID. But instead, somehow RawHID is pre-emptively disabled, and I have no idea as to how :(

**Reply by NicoHood** [reply]
Check that out:
https://github.com/NicoHood/HID/issues/91
